samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs

```diff
@@ -0,0 +1,184 @@
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            return endpoints.Any(e => e.Metadata.GetMetadata<DomainMetadata>()?.Domains.Count > 0);
+        }
+
+        public IReadOnlyList<PolicyNodeEdge> GetEdges(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            // The algorithm here is designed to be preserve the order of the endpoints
+            // while also being relatively simple. Preserving order is important.
+
+            // First, build a dictionary of all of the domains that are included
+            // at this node.
+            //
+            // For now we're just building up the set of keys. We don't add any endpoints
+            // to lists now because we don't want ordering problems.
+            var edges = new Dictionary<HostString, List<Endpoint>>();
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var endpoint = endpoints[i];
+                var domains = endpoint.Metadata.GetMetadata<DomainMetadata>()?.Domains;
+                if (domains == null || domains.Count == 0)
+                {
+                    domains = new[] { new HostString() };
+                }
+
+                for (var j = 0; j < domains.Count; j++)
+                {
+                    var contentType = domains[j];
+
+                    if (!edges.ContainsKey(contentType))
+                    {
+                        edges.Add(contentType, new List<Endpoint>());
+                    }
+                }
+            }
+
+            // Now in a second loop, add endpoints to these lists. We've enumerated all of
+            // the states, so we want to see which states this endpoint matches.
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var endpoint = endpoints[i];
+                var domains = endpoint.Metadata.GetMetadata<DomainMetadata>()?.Domains ?? Array.Empty<HostString>();
+                if (domains.Count == 0)
+                {
+                    // OK this means that this endpoint matches *all* domains.
+                    // So, loop and add it to all states.
+                    foreach (var kvp in edges)
+                    {
+                        kvp.Value.Add(endpoint);
+                    }
+                }
+                else
+                {
+                    // OK this endpoint matches specific domains
+                    foreach (var kvp in edges)
+                    {
+                        // The edgeKey maps to a possible request header value
+                        var edgeKey = kvp.Key;
+
+                        for (var j = 0; j < domains.Count; j++)
+                        {
+                            var domain = domains[j];
+
+                            if (edgeKey == domain)
```

Comment:
Is this logic correct when you have wildcards? It sounds like it's not, but you expect the ordering of the domains to help you. Consider a case like this:

Endpoint1:
domain: foo.example.com
version: v2

Endpoint2:
domain *.example.com
version: v1

This code will be build edges like this:

foo.example.com -> (Endpoint1)
*.example.com -> (Endpoint2)

So what if the request is for foo.example.com - version 1?

Reply:
OR 

> maybe I misunderstand what you're saying about wildcards? I kinda expected to see handling for something like *.example.com